appcues/src/main/java/com/appcues/ui/RequestPermissionActivity.kt

```diff
@@ -0,0 +1,96 @@
+        var completion: CompletableDeferred<Boolean>? = null
+
+        fun getIntent(context: Context, permission: String): Intent = Intent(context, RequestPermissionActivity::class.java)
+            .apply {
+                putExtra(EXTRA_PERMISSION_KEY, permission)
+                flags = Intent.FLAG_ACTIVITY_NEW_TASK
+            }
+    }
+
+    private val permissionKey by lazy { intent.getStringExtra(EXTRA_PERMISSION_KEY)!! }
+
+    private val permissionLauncher = registerForActivityResult(RequestPermission()) {
+        completion?.complete(it)
+        finish()
+    }
+
+    private var settingsLauncher = registerForActivityResult(StartActivityForResult()) {
+        // check permission again and return the completion
+        val hasPermission = ContextCompat.checkSelfPermission(this, permissionKey) == PackageManager.PERMISSION_GRANTED
+        completion?.complete(hasPermission)
+        finish()
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        // remove enter animation from this activity
+        if (VERSION.SDK_INT >= VERSION_CODES.UPSIDE_DOWN_CAKE) {
+            overrideActivityTransition(OVERRIDE_TRANSITION_OPEN, 0, 0)
+        } else {
+            @Suppress("DEPRECATION")
+            overridePendingTransition(0, 0)
+        }
+        super.onCreate(savedInstanceState)
+
+        val hasPermission = ContextCompat.checkSelfPermission(this, permissionKey)
+
+        when {
+            // first we check if the permission is already granted
+            hasPermission == PackageManager.PERMISSION_GRANTED -> {
+                completion?.complete(true)
+                finish()
+            }
+            // POST_NOTIFICATION did not exist before API 33, so instead we direct user to open the settings page
+            VERSION.SDK_INT < VERSION_CODES.TIRAMISU &&
+                permissionKey == Manifest.permission.POST_NOTIFICATIONS &&
+                !NotificationManagerCompat.from(this).areNotificationsEnabled() -> {
+                val intent = Intent().apply {
+                    action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS
+                    data = Uri.fromParts("package", packageName, null)
+                }
+                settingsLauncher.launch(intent)
+            }
+            // Permission is denied, we try to launch the permission launcher,
+            // if user has already declined too many times it will instantly return false
+            hasPermission == PackageManager.PERMISSION_DENIED -> {
+                permissionLauncher.launch(permissionKey)
+            }
+        }
+    }
+
+    override fun finish() {
+        super.finish()
+        // remove exit animation from this activity
+        if (VERSION.SDK_INT >= VERSION_CODES.UPSIDE_DOWN_CAKE) {
+            overrideActivityTransition(OVERRIDE_TRANSITION_CLOSE, 0, 0)
+        } else {
+            @Suppress("DEPRECATION")
+            overridePendingTransition(0, 0)
+        }
+    }
```

Comment:
is there any way that the activity could get into here outside of the expected path? Wondering if there is a way the completion callback would not be invoked.

Reply:
so I tested different apis right.

before 33 you dont actually have the permission, thats why we need to check the API level and instead of showing the permission dialog we send them to the settings page.

both scenarios (requestin permission and settings page) when coming back here will resolve and finish() + complete the task,  the other possible scenario is when they already have permission, or in case permission returns false right away (also handled by the launcher)

I could not think of any other scenarios surrounding this for now

Reply:
what about a deep link coming in and redirecting the app to another activity or something like that? could it leave an action in a hung state?

Reply:
not able to reproduce any issue of this sort but as a safety measure for now i will add a delayedJob that cancels this activity if it stays in resumed state for long (should never happen)